messaging/src/main/java/org/axonframework/eventhandling/deadletter/jdbc/GenericDeadLetterTableFactory.java

```diff
@@ -0,0 +1,110 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.eventhandling.deadletter.jdbc;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+
+/**
+ * A {@link DeadLetterTableFactory} implementation compatible with most databases.
+ *
+ * @author Steven van Beelen
+ * @since 4.8.0
+ */
+@SuppressWarnings({"SqlDialectInspection", "SqlNoDataSourceInspection"})
+public class GenericDeadLetterTableFactory implements DeadLetterTableFactory {
+
+    @SuppressWarnings("SqlNoDataSourceInspection")
+    @Override
+    public PreparedStatement createTable(Connection connection, DeadLetterSchema schema) throws SQLException {
+        String sql = "CREATE TABLE IF NOT EXISTS " + schema.deadLetterTable() + " (\n" +
+                schema.deadLetterIdentifierColumn() + " VARCHAR(255) NOT NULL,\n" +
+                schema.processingGroupColumn() + " VARCHAR(255) NOT NULL,\n" +
+                schema.sequenceIdentifierColumn() + " VARCHAR(255) NOT NULL,\n" +
+                schema.sequenceIndexColumn() + " BIGINT NOT NULL,\n" +
+                schema.messageTypeColumn() + " VARCHAR(255) NOT NULL,\n" +
+                schema.eventIdentifierColumn() + " VARCHAR(255) NOT NULL,\n" +
+                schema.timestampColumn() + " " + timestampType() + " NOT NULL,\n" +
+                schema.payloadTypeColumn() + " VARCHAR(255) NOT NULL,\n" +
+                schema.payloadRevisionColumn() + " VARCHAR(255),\n" +
+                schema.payloadColumn() + " " + serializedDataType() + " NOT NULL,\n" +
+                schema.metaDataColumn() + " " + serializedDataType() + ",\n" +
+                schema.aggregateTypeColumn() + " VARCHAR(255),\n" +
+                schema.aggregateIdentifierColumn() + " VARCHAR(255),\n" +
+                schema.sequenceNumberColumn() + " BIGINT,\n" +
+                schema.tokenTypeColumn() + " VARCHAR(255),\n" +
+                schema.tokenColumn() + " " + serializedDataType() + ",\n" +
+                schema.enqueuedAtColumn() + " " + timestampType() + " NOT NULL,\n" +
+                schema.lastTouchedColumn() + " " + timestampType() + ",\n" +
+                schema.processingStartedColumn() + " " + timestampType() + ",\n" +
+                schema.causeTypeColumn() + " VARCHAR(255),\n" +
+                schema.causeMessageColumn() + " VARCHAR(1023),\n" +
+                schema.diagnosticsColumn() + " " + serializedDataType() + ",\n" +
+                "CONSTRAINT PK PRIMARY KEY (" + schema.deadLetterIdentifierColumn() + "),\n" +
```

Comment:
This definition already contains the indexes, why are there separate functions for it as well?
And this one is better, as the sequence index is unique

Reply:
Also, We have to check whether this statement works with enough databases to warrant it to be generic.

Reply:
Perhaps this is me wrongly translating the `DeadLetterEntry` you wrote, but for sake of argument:

```java
@Entity
@Table(indexes = {
        @Index(columnList = "processingGroup"),
        @Index(columnList = "processingGroup,sequenceIdentifier"),
        @Index(columnList = "processingGroup,sequenceIdentifier,sequenceIndex", unique = true),
})
public class DeadLetterEntry {

    @Id
    private String deadLetterId;
...
}
```

I wanted to have the same as the entry.
However, with JDBC, I can _only_ provide uniqueness constraints during the table creation on a generic SQL level.
As only the primary key (the `deadLetterIdentifierColumn`) and the processingGroup, sequenceIdentifier, and sequenceIndex are unique, those are the only specifics I could add to the `createTable` operation.
So, the other indices, for the `processingGroup` and `[processingGroup, sequenceIdentifier]` indices _had_ to be separate operations.

Searching the web, I did find some databases allowing construction of non-unique indices in a `CREATE TABLE` statement.
However, then we wouldn't be dealing with a `GenericDeadLetterTableFactory` anymore, of course.

Perhaps my SQL-knowledge is lacking here, though.
So if you know of a generic SQL working for all databases that also includes non-unique indices, I am all for a suggestion!

Reply:
Alright makes sense. But why not create all indices in this statement?